sdks/python/apache_beam/io/gcp/bigtableio_test.py

```diff
@@ -0,0 +1,218 @@
+                                   self.table_id)
+    current_size = 805306368
+    desired_bundle_size = 402653184
+    start_key = b'beam_key0672496'
+    end_key = b'beam_key1582279'
+    count_all = len(list(self.bigtable.range_split_fraction(current_size,
+                                                            desired_bundle_size,
+                                                            start_key,
+                                                            end_key)))
+    desired_bundle_count = current_size/desired_bundle_size
+    self.assertEqual(count_all, desired_bundle_count)
+
+  # Split Range Sized Subranges
+  @mock.patch.object(BigTableSource, 'get_sample_row_keys')
+  def test_split_rage_sized_subranges(self, mock_my_method):
+    mock_my_method.return_value = self.sample_row_keys()
+    self.bigtable = BigTableSource(self.project_id, self.instance_id,
+                                   self.table_id)
+    sample_size_bytes = 805306368
+    desired_bundle_size = 402653184
+    start_key = b'beam_key0672496'
+    end_key = b'beam_key1582279'
+    ranges = self.bigtable.get_range_tracker(start_key, end_key)
+    range_subranges = self.bigtable.split_range_subranges(
+        sample_size_bytes,
+        desired_bundle_size,
+        ranges)
+    count_all = len(list(range_subranges))
+    desired_bundle_count = sample_size_bytes/desired_bundle_size
+    self.assertEqual(count_all, desired_bundle_count)
+
+  # Range Tracker
+  @mock.patch.object(BigTableSource, 'get_sample_row_keys')
+  def test_get_range_tracker(self, mock_my_method):
+    mock_my_method.return_value = self.sample_row_keys()
+    self.bigtable = BigTableSource(self.project_id, self.instance_id,
+                                   self.table_id)
+    start_position = b'beam_key0672496'
+    stop_position = b'beam_key1582279'
+    range_tracker = self.bigtable.get_range_tracker(
+        start_position,
+        stop_position)
+
+    self.assertEqual(range_tracker.start_position(), start_position)
+    self.assertEqual(range_tracker.stop_position(), stop_position)
+
+  @mock.patch.object(BigTableSource, 'get_sample_row_keys')
+  def test_split_four_hundred_bytes(self, mock_my_method):
+    mock_my_method.return_value = self.sample_row_keys()
+    bigtable = BigTableSource(self.project_id, self.instance_id,
+                              self.table_id)
+    desired_bundle_size = 402653184
+    split = bigtable.split(desired_bundle_size)
+    split_list = list(split)
+    count = len(split_list)
+    size = 10468982784
+    count_size = size/desired_bundle_size
+    count_size += 1
+    self.assertEqual(count, count_size)
+
+  def __read_list(self):
+    for i in range(72496, 72500):
+      key = 'beam_key'+str(i)
+      if sys.version_info < (3, 0):
+        key = bytes(key)
+      else:
+        key = bytes(key, 'utf8')
+      yield PartialRowData(key)
+
+  @mock.patch.object(Table, 'read_rows')
+  def test_read(self, mock_read_rows):
```

Comment:
Can we add a exhaustive test for dynamic work rebalancing similar to following (using mocks for BT) ?
https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/textio_test.py#L386

Reply:
Yes we can do it, but, i need to figure out what to mock in bigtable.

Reply:
About this test, I getting some errors when the assert use the try_split with `0.0` fraction and get the position, and when trying to get the try_split result of the bundle `[b'', b'beam_key2')`, and set the `b''`(first element) and get a `ValueError: Split at '' not in range ['', 'beam_key2']` error.
I don't know why I getting that error